cerberus/models/booking.py

```diff
@@ -187,47 +286,99 @@ def save(self, *args, **kwargs) -> None:
+
+    def check_valid(self) -> None:
+        if any(pet.customer != self.customer for pet in self.pets.all()):
+            raise ValidationError("Booking has pets from a different customer")
+
+    def natural_date(self):
+        return naturaldate(self.start)
+
+    @classmethod
+    def get_mix_max_time(cls, date: date) -> tuple[datetime | None, datetime | None]:
+        date = make_aware(datetime(date.year, date.month, date.day))
+        next_date = date + timedelta(days=1)
+
+        result = (
+            cls.objects.filter(start__gt=date, end__lt=next_date).values("start").aggregate(Min("start"), Max("end"))
+        )
+
+        return result["start__min"], result["end__max"]
+
+    @property
+    def name(self) -> str:
+        pet_names = ", ".join(str(p) for p in self.pets.all())
+        return f"{pet_names} - {self.service}"
 
     @property
     def booking_slot(self) -> BookingSlot:
         if self._booking_slot is None:
-            if self.state == self.States.CANCELED.value:
+            if self.state == BookingStates.CANCELED.value:
                 raise (BookingSlot.DoesNotExist("Booking has been canceled"))
             self._booking_slot = self._get_new_booking_slot()
         return self._booking_slot
 
     @booking_slot.setter
-    def booking_slot(self, value: BookingSlot) -> None:
-        self._booking_slot = value
+    def booking_slot(self, value: BookingSlot | None) -> None:
+        if value != self._booking_slot:
+            self._previous_slot = self._booking_slot
 
-    def create_charge(self) -> Charge:
-        charge = BookingCharge(
-            name=f"Charge for {self.name}"[:255],
-            line=self.cost,
-            booking=self,
-            customer=self.pet.customer,
-        )
-        charge.save()
+        self._booking_slot = value
 
-        return charge
+    def create_charges(self) -> list[Charge]:
+        cost = self.cost
+        if self.cost_per_additional is None:
+            pet_names = ", ".join(str(p) for p in self.pets.all())
+            name = f"{self.service} for {pet_names}"[:255]
+            charge = BookingCharge(name=name, line=cost, booking=self, customer=self.customer)
+            charge.save()
+            return [charge]
+
+        charges = []
+
+        if self.pets.count():
+            for pet in self.pets.all():
+                name = f"{self.service} for {pet}"[:255]
+                charge = BookingCharge(name=name, line=cost, booking=self, customer=self.customer)
+                charge.save()
+                charges.append(charge)
+
+                cost = self.cost_per_additional
+                if self.cost_per_additional is None:
+                    break
+        else:
+            name = f"{self.service}"[:255]
+            charge = BookingCharge(name=name, line=cost, booking=self, customer=self.customer)
+            charge.save()
+            charges.append(charge)
+
+        return charges
```

Comment:
<a href="https://sourcery.ai"><img src="https://sourcery.ai/favicon.svg" height="15" alt="Sourcery logo" /></a>  **suggestion (performance):** Consider bulk creation of `Charge` objects.

Creating `Charge` objects in a loop might not be the most efficient approach. Explore using `bulk_create` for better performance.